### PR TITLE
Run a call back when the session token is expired

### PR DESCRIPTION
**Pull request by blfakir**
## Description
Run a callBack when the session token expire.
const connectionParameters = sdk.ConnectionParameters.ofBearerToken("http://acc-sdk:8080", 
                                                    "$token$", {onSessionExpiration: ()=> console.log("session token expired")});

## How Has This Been Tested?
Unit tests
Manual tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -127,6 +127,7 @@ traceAPICalls|false| Activates tracing of API calls or not
 transport|axios|Overrides the transport layer
 noStorage|false|De-activate using of local storage
 storage|localStorage|Overrides the local storage for caches
+onSessionExpiration|undefined|Callback to run when the session token is expired
```
Comment (mkiki) [inline README.md:130]: We need a section in the readme which explain how to use and gives an example. For instance, what is supposed to be implemented in the callback. Should the callback call logon again? Should it return anything?

File: src/client.js [hunk 2]
```diff
@@ -707,6 +711,9 @@ class Client {
                 if (that._traceAPICalls)
                     console.log(`SOAP//failure ${ex.toString()}`);
                 that._notifyObservers((observer) => observer.onSOAPCallFailure && observer.onSOAPCallFailure(soapCall, ex) );
+                // Call session expiration callback in case of XSV-350008 Campaign error (Session expiration)
+                if (ex.errorCode == 'XSV-350008')
```
Comment (mkiki) [inline src/client.js:715]: You can use "==="

File: src/soap.js [hunk 3]
```diff
@@ -565,6 +565,22 @@ class SoapMethodCall {
         const that = this;
         const promise = this._transport(this.request);
         return promise.then(function(body) {
+            // Return a XML SOAP response in case of session expiration error
```
Comment (mkiki) [inline src/soap.js:568]: I do not understand this. Shouldn't the session expiration have thrown an exception?

File: test/client.test.js [hunk 4]
```diff
@@ -1899,7 +1899,7 @@ describe('ACC Client', function () {
     describe("Bearer token authentication", () => {
         // Bearer token authentication is used when embedding IMS for authentication
         it("Should create logged client", async() => {
-            const connectionParameters = sdk.ConnectionParameters.ofBearerToken("http://acc-sdk:8080", "$token$");
```
Comment (mkiki) [inline test/client.test.js:1902]: Is this change related to this PR? I'm not sure I understand it: why calling ofBearerToken without a bearer token?

File: test/client.test.js [hunk 5]
```diff
@@ -1946,6 +1946,47 @@ describe('ACC Client', function () {
...
+                }
+            };
+
+            client._transport.mockReturnValueOnce(Promise.resolve(`XSV-350008 Session has expired or is invalid. Please reconnect.`));
+
+            var query = client.NLWS.xtkQueryDef.create(queryDef);
+            await expect(query.executeQuery()).rejects.toMatchObject({ errorCode: "XSV-350008" });
+            expect(onSessionExpiration.mock.calls.length).toBe(1);
+            
+            client._transport.mockReturnValueOnce(Promise.resolve(`<?xml version='1.0'?>
```
Comment (mkiki) [inline test/client.test.js:1976]: I think it would be useful to illustrate how to log on again after we have a new bearer token

**Review by mkiki — CHANGES_REQUESTED** [review]
Hi Badre, please see comments in the review.
Alos, 1.0.6 has been released, so we need 1.0.7 now ;-)

File: README.md (hunk 1 above)
Comment (blfakir) [reply README.md:130]: done

File: src/client.js (hunk 2 above)
Comment (blfakir) [reply src/client.js]: I did a change to use the statusCode instead of the error code

File: src/soap.js (hunk 3 above)
Comment (blfakir) [reply src/soap.js]: Done thanks for the hint

File: test/client.test.js (hunk 4 above)
Comment (blfakir) [reply test/client.test.js:1902]: This is only for code coverage

File: test/client.test.js (hunk 5 above)
Comment (blfakir) [reply test/client.test.js:1980]: I added an example login gain with the same credential. I think we need to  modify the client in order to support the use a new bearer token with the same client.

File: README.md [hunk 6]
```diff
@@ -219,6 +220,19 @@ await client.logon();
 await client.logoff();
 ```
 
+## onSessionExpiration callback
+The onSessionExpiration is callback with the sdk client as parameter, it is called when the ACC session is expired.
```
Comment (mkiki) [inline README.md:224]: The onSessionExpiration is **a** callback with the sdk client as parameter, it is called when the ACC session is expired.

File: README.md [hunk 7]
```diff
@@ -219,6 +220,19 @@ await client.logon();
...
+The callback return no thing. We can use the callback to reconnect the sdk client.
```
Comment (mkiki) [inline README.md:225]: The callback return **nothing**. We can use the callback to reconnect the sdk client.

File: README.md [hunk 8]
```diff
@@ -219,6 +220,19 @@ await client.logon();
...
+
+```js
+const connectionParameters = sdk.ConnectionParameters.ofUserAndPassword(
+                                url, "admin", "admin",
+                                {onSessionExpiration: async (client)=>{
+                                        await client.logon();
```
Comment (mkiki) [inline README.md:231]: What if logon returns an expired session again? How do we prevent infinite loops?

File: src/client.js [hunk 9]
```diff
@@ -707,6 +711,9 @@ class Client {
                 if (that._traceAPICalls)
                     console.log(`SOAP//failure ${ex.toString()}`);
                 that._notifyObservers((observer) => observer.onSOAPCallFailure && observer.onSOAPCallFailure(soapCall, ex) );
+                // Call session expiration callback in case of 403
+                if (ex.statusCode == 403)
```
Comment (mkiki) [inline src/client.js:715]: Why? 403 is forbidden. It may not be a session expiration. Would there not be an infinite loop in this xase?

Comment (mkiki) [reply src/client.js:715]: I'm not sure of the logic here.
The callback is asynchronous, shouldn't there be an await here?
Are we retrying the failed call after a failure?

**Review by mkiki — CHANGES_REQUESTED** [review]
